### ISSUE TITLE
Using inference_program meets problem

### ISSUE THREAD

**Issue by BigFishMaster** [PaddlePaddle/Paddle#9334]
ENV:
```
cuda9
cudnn7
Machine: V100, 8 cards
```
I use ```inference_program``` in two ways in my code. The first way is to clone ```default_main_program``` before the optimization. The other one is to use ```get_inference_program```. 
However, the first way can output right validation accuracy but the second one will output random validation accuracy.
I test this code using 10 classes from ImageNet1000. The wrong second way always output ~10% top-1 accuracy after each epoch.

Details are below:
One:
```
   inference_program = fluid.default_main_program().clone()
```
The other:
```
   inference_program = fluid.default_main_program().clone()
   with fluid.program_guard(inference_program):
       inference_program = fluid.io.get_inference_program([avg_cost, acc_top1, acc_top5])
```
My code is as follows:
```
    image = fluid.layers.data(name='image', shape=image_shape, dtype='float32')
    label = fluid.layers.data(name='label', shape=[1], dtype='int64')

    if parallel:
        places = fluid.layers.get_places()
        pd = fluid.layers.ParallelDo(places, use_nccl=use_nccl)

        with pd.do():
            image_ = pd.read_input(image)
            label_ = pd.read_input(label)
            out = SE_ResNeXt(input=image_, class_dim=class_dim, layers = layers)
            cost = fluid.layers.cross_entropy(input=out, label=label_)
            avg_cost = fluid.layers.mean(x=cost)
            acc_top1 = fluid.layers.accuracy(input=out, label=label_, k=1)
            acc_top5 = fluid.layers.accuracy(input=out, label=label_, k=5)
            pd.write_output(avg_cost)
            pd.write_output(acc_top1)
            pd.write_output(acc_top5)

        avg_cost, acc_top1, acc_top5 = pd()
        avg_cost = fluid.layers.mean(x=avg_cost)
        acc_top1 = fluid.layers.mean(x=acc_top1)
        acc_top5 = fluid.layers.mean(x=acc_top5)
    else:
        out = SE_ResNeXt(input=image, class_dim=class_dim, layers = layers)
        cost = fluid.layers.cross_entropy(input=out, label=label)
        avg_cost = fluid.layers.mean(x=cost)
        acc_top1 = fluid.layers.accuracy(input=out, label=label, k=1)
        acc_top5 = fluid.layers.accuracy(input=out, label=label, k=5)
   # inference_program = fluid.default_main_program().clone()

    if lr_strategy is None:
        optimizer = fluid.optimizer.Momentum(
            learning_rate=learning_rate,
            momentum=0.9,
            regularization=fluid.regularizer.L2Decay(1e-4))
    else:
        bd = lr_strategy["bd"]
        lr = lr_strategy["lr"]
        optimizer = fluid.optimizer.Momentum(
            learning_rate=fluid.layers.piecewise_decay(
                boundaries=bd, values=lr),
            momentum=0.9,
            regularization=fluid.regularizer.L2Decay(1e-4))

    opts = optimizer.minimize(avg_cost)
    fluid.memory_optimize(fluid.default_main_program())

    #inference_program = fluid.default_main_program().clone()
    #with fluid.program_guard(inference_program):
    #    inference_program = fluid.io.get_inference_program([avg_cost, acc_top1, acc_top5])
```

**Comment by BigFishMaster** [comment]
logs on 10-class classification:
```
Pass 0, trainbatch 0, loss 2.71465396881, acc1 0.121031746268, acc5 0.539285778999 time 4.02 sec
Pass 0, trainbatch 10, loss 2.75805997849, acc1 0.121428579092, acc5 0.464285731316 time 1.23 sec
Pass 0, testbatch 10, loss 56.660320282, acc1 0.114682555199, acc5 0.510317504406 time 0.57 sec
Pass 0, testbatch 10, loss 58.7159729004, acc1 0.0825630277395, acc5 0.487394988537 time 0.67 sec
End pass 0, train_loss 6.02023363113, train_acc1 0.141847059131, train_acc5 0.514105379581,              test_
loss 57.6881484985, test_acc1 0.0986227914691, test_acc5 0.498856246471
Pass 1, trainbatch 0, loss 2.62622570992, acc1 0.143650814891, acc5 0.510317504406 time 1.38 sec
Pass 1, trainbatch 10, loss 2.33206248283, acc1 0.116269849241, acc5 0.525396823883 time 1.29 sec
Pass 1, testbatch 10, loss 56.660320282, acc1 0.114682555199, acc5 0.510714352131 time 0.82 sec
Pass 1, testbatch 10, loss 58.7159729004, acc1 0.0825630277395, acc5 0.486764758825 time 0.70 sec
End pass 1, train_loss 3.21767783165, train_acc1 0.118578664958, train_acc5 0.511327624321,              test_
loss 57.6881484985, test_acc1 0.0986227914691, test_acc5 0.498739540577
Pass 2, trainbatch 0, loss 2.61714243889, acc1 0.136507943273, acc5 0.558730125427 time 1.81 sec
Pass 2, trainbatch 10, loss 2.34106326103, acc1 0.121428579092, acc5 0.555555582047 time 1.56 sec
Pass 2, testbatch 10, loss 58.7428627014, acc1 0.0821428596973, acc5 0.513888955116 time 1.04 sec
Pass 2, testbatch 10, loss 56.2016792297, acc1 0.121848739684, acc5 0.483193308115 time 0.89 sec
End pass 2, train_loss 2.59958982468, train_acc1 0.143506512046, train_acc5 0.565945208073,              test_
loss 57.4722709656, test_acc1 0.101995795965, test_acc5 0.498541116714
Pass 3, trainbatch 0, loss 2.41077566147, acc1 0.162698417902, acc5 0.577381014824 time 1.69 sec
Pass 3, trainbatch 10, loss 2.20666956902, acc1 0.140079364181, acc5 0.643254041672 time 1.79 sec
Pass 3, testbatch 10, loss 56.660320282, acc1 0.114682555199, acc5 0.510317504406 time 1.24 sec
Pass 3, testbatch 10, loss 58.7159729004, acc1 0.0825630277395, acc5 0.487394958735 time 1.10 sec
End pass 3, train_loss 2.33437418938, train_acc1 0.198268398643, train_acc5 0.642460286617,              test_
loss 57.6881484985, test_acc1 0.0986227914691, test_acc5 0.498856246471
Pass 4, trainbatch 0, loss 2.42051196098, acc1 0.190873011947, acc5 0.627381026745 time 2.07 sec
Pass 4, trainbatch 10, loss 2.13302111626, acc1 0.177777782083, acc5 0.651587367058 time 1.92 sec
Pass 4, testbatch 10, loss 56.660320282, acc1 0.114682555199, acc5 0.510317504406 time 1.29 sec
Pass 4, testbatch 10, loss 58.7159729004, acc1 0.0825630277395, acc5 0.486764758825 time 1.17 sec
End pass 4, train_loss 2.23031830788, train_acc1 0.210425674915, train_acc5 0.727200567722,              test_
loss 57.6881484985, test_acc1 0.0986227914691, test_acc5 0.498541116714
```
Logs on 1000-class classification
```
Pass 1, trainbatch 4960, loss 4.22036886215, acc1 0.262301594019, acc5 0.41746032238 time 1.22 sec
Pass 1, trainbatch 4970, loss 3.79807925224, acc1 0.27380952239, acc5 0.452380955219 time 1.26 sec
Pass 1, trainbatch 4980, loss 3.73623943329, acc1 0.305158734322, acc5 0.497222274542 time 1.23 sec
Pass 1, trainbatch 4990, loss 3.01816082001, acc1 0.359126985073, acc5 0.588888883591 time 1.21 sec
Pass 1, trainbatch 5000, loss 4.31636953354, acc1 0.230555549264, acc5 0.425000011921 time 1.24 sec
End pass 1, train_loss 4.2414727211, train_acc1 0.217968136072, train_acc5 0.408199131489,              test_l
oss 63.737991333, test_acc1 0.00241537112743, test_acc5 0.00699506001547
Pass 2, trainbatch 0, loss 3.58182191849, acc1 0.276190489531, acc5 0.521031796932 time 1.45 sec
Pass 2, trainbatch 10, loss 3.42263197899, acc1 0.321428596973, acc5 0.547619104385 time 1.58 sec
Pass 2, trainbatch 20, loss 3.36182427406, acc1 0.291666686535, acc5 0.553571462631 time 1.17 sec
Pass 2, trainbatch 30, loss 3.81796765327, acc1 0.266666710377, acc5 0.4623016119 time 1.18 sec
Pass 2, trainbatch 40, loss 3.30461668968, acc1 0.335714280605, acc5 0.533730208874 time 1.23 sec
Pass 2, trainbatch 50, loss 4.86434936523, acc1 0.203174620867, acc5 0.326190471649 time 1.23 sec
...
Pass 2, trainbatch 4930, loss 5.21048879623, acc1 0.182539686561, acc5 0.305555552244 time 1.21 sec
Pass 2, trainbatch 4940, loss 3.05516839027, acc1 0.328571438789, acc5 0.607142865658 time 1.26 sec
Pass 2, trainbatch 4950, loss 3.74800610542, acc1 0.274603188038, acc5 0.485714286566 time 1.22 sec
Pass 2, trainbatch 4960, loss 3.16665887833, acc1 0.334127038717, acc5 0.595634937286 time 1.22 sec
Pass 2, trainbatch 4970, loss 3.44320034981, acc1 0.329365104437, acc5 0.507936537266 time 1.25 sec
Pass 2, trainbatch 4980, loss 4.53301334381, acc1 0.22619047761, acc5 0.380555570126 time 1.21 sec
Pass 2, trainbatch 4990, loss 3.46615171432, acc1 0.349603205919, acc5 0.548412740231 time 1.23 sec
Pass 2, trainbatch 5000, loss 3.23511362076, acc1 0.33373016119, acc5 0.603571534157 time 1.25 sec
End pass 2, train_loss 3.79994702339, train_acc1 0.287314504385, train_acc5 0.489271908998,              test_
loss 62.0461463928, test_acc1 0.000931325019337, test_acc5 0.00889415293932
Pass 3, trainbatch 0, loss 3.06240296364, acc1 0.392460376024, acc5 0.619047641754 time 1.22 sec
Pass 3, trainbatch 10, loss 3.81386899948, acc1 0.285317480564, acc5 0.50357145071 time 1.19 sec
Pass 3, trainbatch 20, loss 4.75208091736, acc1 0.194047629833, acc5 0.340873062611 time 1.21 sec
Pass 3, trainbatch 30, loss 5.21858882904, acc1 0.138888895512, acc5 0.285714298487 time 1.22 sec
```

**Comment by Xreki** [comment]
It is strange... I think the first usage should be wrong...

**Comment by BigFishMaster** [comment]
@Xreki Why do you think it is wrong?

**Comment by BigFishMaster** [comment]
Single card logs using second way. I think it is ok...
```
Pass 0, trainbatch 0, loss 7.01783514023, acc1 0.019999999553, acc5 0.0399999991059 time 1.35 sec
Pass 0, trainbatch 10, loss 9.53971672058, acc1 0.0599999986589, acc5 0.479999989271 time 1.24 sec
Pass 0, trainbatch 20, loss 4.33643293381, acc1 0.140000000596, acc5 0.560000002384 time 1.22 sec
Pass 0, trainbatch 30, loss 6.04275560379, acc1 0.019999999553, acc5 0.479999989271 time 1.32 sec
Pass 0, trainbatch 40, loss 3.61183261871, acc1 0.0599999986589, acc5 0.519999980927 time 1.19 sec
Pass 0, trainbatch 50, loss 2.95087981224, acc1 0.0799999982119, acc5 0.5 time 1.19 sec
Pass 0, testbatch 0, loss 2.86973881721, acc1 0.140000000596, acc5 0.539999961853 time 0.62 sec
End pass 0, train_loss 6.16225481033, train_acc1 0.11215686053, train_acc5 0.531764626503,              test_loss 3.18870210648, test_acc1 0.155999988317, test_acc5 0.6099999547
Pass 1, trainbatch 0, loss 3.18502759933, acc1 0.119999997318, acc5 0.519999980927 time 1.81 sec
Pass 1, trainbatch 10, loss 2.42919206619, acc1 0.0799999982119, acc5 0.539999961853 time 1.81 sec
Pass 1, trainbatch 20, loss 2.33842587471, acc1 0.119999997318, acc5 0.620000004768 time 2.47 sec
Pass 1, trainbatch 30, loss 2.25649523735, acc1 0.199999988079, acc5 0.680000007153 time 1.94 sec
Pass 1, trainbatch 40, loss 2.3714761734, acc1 0.179999992251, acc5 0.740000009537 time 2.09 sec
Pass 1, trainbatch 50, loss 1.92521083355, acc1 0.319999992847, acc5 0.77999997139 time 2.16 sec
Pass 1, testbatch 0, loss 2.06457877159, acc1 0.179999992251, acc5 0.8599999547 time 1.26 sec
End pass 1, train_loss 2.80949163437, train_acc1 0.147843137383, train_acc5 0.623137295246,              test_loss 2.08758378029, test_acc1 0.201999992132, test_acc5 0.804000020027
Pass 2, trainbatch 0, loss 4.33151197433, acc1 0.159999996424, acc5 0.5 time 2.78 sec
Pass 2, trainbatch 10, loss 1.96619009972, acc1 0.179999992251, acc5 0.799999952316 time 3.12 sec
```

**Comment by BigFishMaster** [comment]
@Xreki I use ```for_test=True```， the test performance is very pool. I think there are bugs in the code.  
Please make full test... I have meet too many bugs this week
```
Pass 1, testbatch 210, loss 63.6815071106, acc1 0.00510204117745, acc5 0.00510204117745 time 0.81 sec
Pass 1, testbatch 220, loss 64.0100021362, acc1 0.0, acc5 0.0102040823549 time 0.81 sec
Pass 1, testbatch 230, loss 63.401473999, acc1 0.00956632662565, acc5 0.00956632662565 time 0.82 sec
Pass 1, testbatch 240, loss 63.3565101624, acc1 0.00510204117745, acc5 0.0102040823549 time 0.81 sec
End pass 1, train_loss 4.27716302872, train_acc1 0.212268814445, train_acc5 0.402588039637,              test_
loss 63.8389282227, test_acc1 0.00220408174209, test_acc5 0.00555867375806
Pass 2, trainbatch 0, loss 4.29365730286, acc1 0.244897991419, acc5 0.418367415667 time 2.97 sec
Pass 2, trainbatch 10, loss 4.18189907074, acc1 0.258290857077, acc5 0.442602068186 time 3.25 sec
Pass 2, trainbatch 20, loss 3.08782553673, acc1 0.374362289906, acc5 0.589923501015 time 2.93 sec
Pass 2, trainbatch 30, loss 4.85054636002, acc1 0.209183678031, acc5 0.341199010611 time 2.96 sec
Pass 2, trainbatch 40, loss 3.34561514854, acc1 0.29719388485, acc5 0.487882703543 time 2.91 sec
```

**Comment by qingqing01** [comment]
Related to #9386
